tokenizers/src/models/unigram/model.rs

```diff
@@ -0,0 +1,433 @@
+        let other_vocab: Vec<(&String, &f64)> =
+            other.vocab.iter().zip(other.scores.iter()).collect();
+        self.unk_id == other.unk_id && vocab == other_vocab
+    }
+}
+
+impl std::fmt::Debug for Unigram {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct("BPE")
+            .field("vocab", &self.vocab.len())
+            .finish()
+    }
+}
+
+static K_UNK_PENALTY: f64 = 10.0;
+
+impl Default for Unigram {
+    fn default() -> Self {
+        let vocab = vec![("<unk>".to_string(), 0.0)];
+        Self::from(&vocab, 0)
+    }
+}
+
+impl Unigram {
+    /// Create a `Unigram` model from a given vocabulary.
+    /// Vocabulary are the various tokens and their associated score which is a sort of a logprob of
+    /// their frequency, which will enable tokenization and sampling.
+    /// unk_id, is the index within the vocabulary.
+    /// For now `Unigram` *requires* at least `unk` because we might find a never seen char.
+    /// Further versions might allow that part to be hidden.
+    pub fn from(vocabulary: &[(String, f64)], unk_id: usize) -> Self {
+        let n = vocabulary.len();
+        let mut vocab: Vec<String> = Vec::with_capacity(n);
+        let mut scores: Vec<f64> = Vec::with_capacity(n);
+        let mut token_to_ids: TokenMap = HashMap::new();
+        let mut builder = TrieBuilder::default();
+
+        assert!(n >= 1, "We need at least unk in the vocabulary");
+        assert!(unk_id < vocabulary.len(), "Unk id is invalid");
+
+        let bos_id = n + 1;
+        let eos_id = n + 2;
+
+        for (id, (token, score)) in vocabulary.iter().enumerate() {
+            vocab.push(token.to_string());
+            scores.push(*score);
+            token_to_ids.insert(token.to_string(), id as u32);
+            let chars: Vec<char> = token.chars().collect();
+            builder.push(&chars);
+        }
+        let min_score = scores.iter().fold(f64::INFINITY, |a, &b| a.min(b));
+        if min_score == -f64::INFINITY {
+            panic!("Alert min_score !!");
+        }
+        let trie = builder.build();
+        let fuse_unk = true;
+
+        Unigram {
+            vocab,
+            scores,
+            token_to_ids,
+            trie,
+            min_score,
+            bos_id,
+            eos_id,
+            unk_id,
+            fuse_unk,
+        }
+    }
+
+    #[cfg(test)]
+    pub(super) fn set_fuse_unk(&mut self, fuse_unk: bool) {
+        self.fuse_unk = fuse_unk;
+    }
+
+    pub(super) fn len(&self) -> usize {
+        self.vocab.len()
+    }
+
+    pub(super) fn populate_nodes(&self, lattice: &mut Lattice) {
+        let unk_score = self.min_score - K_UNK_PENALTY;
+
+        let len = lattice.len();
+
+        for begin_pos in 0..len {
+            // let now = Instant::now();
+            let trie_results: Vec<Vec<char>> =
+                self.trie.common_prefix_search(&lattice.chars[begin_pos..]);
```

Comment:
Nit: We could maybe have `common_prefix_search` return a `Vec<String>` directly since it seems to be used only here. Counting chars afterward is probably faster than re-allocating.

Reply:
I'll try